### ISSUE TITLE
BOpt_DestroyMenuEntry function attempts to free NULL in two places (Bugzilla Bug 135)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#6650]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 135
Date: 2016-10-04T21:40:49+00:00
From: Eric Nelson &lt;<eric.nelson>&gt;
To:   Hao Wu &lt;<hao.a.wu>&gt;
CC:   hot.tian, yonghong.zhu

Last updated: 2016-10-19T22:14:13+00:00

**Comment by tianocore-issues** [comment]
## Comment 554
Date: 2016-10-04 21:40:49 +0000
From: Eric Nelson &lt;<eric.nelson>&gt;

- Industry Specification: ---
- Release Observed: UDK 2015
- Releases to Fix: UDK 2015
- Target OS: ---
- Bugzilla Assignee(s): Hao Wu &lt;<hao.a.wu>&gt;

After I build run the r18552 (9/29/15) UDK2015 NT32 IA32 environment, and attempt to add an item to the boot menu (Boot Maintenance Manager =&gt; Boot Options =&gt; Add Boot Option), SecMain.exe crashes:

ASSERT_EFI_ERROR (Status = Invalid Parameter)

ASSERT!: c:\myworkspace\MdePkg\Library\UefiMemoryAllocationLib\MemoryAllocationLib.c (819): !EFI_ERROR (Status)

I have found that the BOpt_DestroyMenuEntry function attempts to free NULL in two places (refer to the file comparison below).

I am using VS2013x86 tool-chain, on Win10 RS1 64-bit OS, and hit this issue in both RELEASE and DEBUG IA32 builds.

~~~~~~ diff of fixes ~~~~~~

Comparing files BootOption.c and BOOTOPTION.NEW.TXT
***** BootOption.c
    if (!FileContext-&gt;IsRoot) {
      FreePool(FileContext-&gt;DevicePath);
    } else {
***** BOOTOPTION.NEW.TXT
    if (!FileContext-&gt;IsRoot) {
                if (FileContext-&gt;DevicePath != NULL) { // Fix #1
                        FreePool(FileContext-&gt;DevicePath);
                }
    } else {
*****

***** BootOption.c

  FreePool(MenuEntry-&gt;DisplayString);
  if (MenuEntry-&gt;HelpString != NULL) {
***** BOOTOPTION.NEW.TXT

  if (MenuEntry-&gt;DisplayString != NULL) { //  // Fix #2
          FreePool(MenuEntry-&gt;DisplayString);
  }
  if (MenuEntry-&gt;HelpString != NULL) {
*****

**Comment by tianocore-issues** [comment]
## Comment 596
Date: 2016-10-13 20:09:14 +0000
From: Yonghong Zhu &lt;<yonghong.zhu>&gt;

Bug scrub: Assign to module owner

**Comment by tianocore-issues** [comment]
## Comment 606
Date: 2016-10-16 23:15:47 +0000
From: Hao Wu &lt;<hao.a.wu>&gt;

(In reply to Eric Nelson from comment #0)
&gt; After I build run the r18552 (9/29/15) UDK2015 NT32 IA32 environment, and
&gt; attempt to add an item to the boot menu (Boot Maintenance Manager =&gt; Boot
&gt; Options =&gt; Add Boot Option), SecMain.exe crashes:
&gt; 
&gt; ASSERT_EFI_ERROR (Status = Invalid Parameter)
&gt; 
&gt; ASSERT!:
&gt; c:\myworkspace\MdePkg\Library\UefiMemoryAllocationLib\MemoryAllocationLib.c
&gt; (819): !EFI_ERROR (Status)
&gt; 
&gt; I have found that the BOpt_DestroyMenuEntry function attempts to free NULL
&gt; in two places (refer to the file comparison below).
&gt; 
&gt; I am using VS2013x86 tool-chain, on Win10 RS1 64-bit OS, and hit this issue
&gt; in both RELEASE and DEBUG IA32 builds.
&gt; 
&gt; ~~~~~~ diff of fixes ~~~~~~
&gt; 
&gt; Comparing files BootOption.c and BOOTOPTION.NEW.TXT
&gt; ***** BootOption.c
&gt;     if (!FileContext-&gt;IsRoot) {
&gt;       FreePool(FileContext-&gt;DevicePath);
&gt;     } else {
&gt; ***** BOOTOPTION.NEW.TXT
&gt;     if (!FileContext-&gt;IsRoot) {
&gt;                 if (FileContext-&gt;DevicePath != NULL) { // Fix #1
&gt;                         FreePool(FileContext-&gt;DevicePath);
&gt;                 }
&gt;     } else {
&gt; *****
&gt; 
&gt; ***** BootOption.c
&gt; 
&gt;   FreePool(MenuEntry-&gt;DisplayString);
&gt;   if (MenuEntry-&gt;HelpString != NULL) {
&gt; ***** BOOTOPTION.NEW.TXT
&gt; 
&gt;   if (MenuEntry-&gt;DisplayString != NULL) { //  // Fix #2
&gt;           FreePool(MenuEntry-&gt;DisplayString);
&gt;   }
&gt;   if (MenuEntry-&gt;HelpString != NULL) {
&gt; *****

Hi Eric,

I am trying to reproduce the above issue using the UDK2015 code rev. 18552 on a Win7 64-bit environment and using VS2013x86 tool-chain DEBUG build. However, I cannot reproduce the issue on my side.

Did you meet the ASSERT when you select the &quot;Add Boot Option&quot; for the first time? Was there any additional changes made beforehand?

Thanks in advance for your help.

**Comment by tianocore-issues** [comment]
## Comment 612
Date: 2016-10-17 15:38:02 +0000
From: Eric Nelson &lt;<eric.nelson>&gt;


No, there were no additional changes.

Furthermore, I first hit this issue on my IT-managed Win10 laptop (HP), and then right before I filed the sighting, I also downloaded it onto a Fresh install of RS1+VS2013 (I was updating from 9600 my Dell XPS 15&quot;, and clean installed it), and again I hit this immediately.

After I launch the --NT32 environment for the first time, after it copies the target.txt to conf, I edited the toolchain (to VS2013x86), and also set the number of threads to one plus the number of cores, so I use 5 on my quad-core i7, whereas I used 3 on my i5 system.

No other changes.

When I try t &quot;add a boot entry&quot; from the boot maintenance manager, it attempts to free those two NULL pointers every time.

It&apos;s a mystery why you don&apos;t hit it, as I am two for two.

Wy not just add the NULL checks, like every other NULL check in that function.  Apparenty there is some case tat can be hit.  Maybe on your system, it is just dumb luck that the uninitialized pointers are set to NULL, whereas on my system(s), they contain garbage.

**Comment by tianocore-issues** [comment]
## Comment 613
Date: 2016-10-17 21:14:39 +0000
From: Hao Wu &lt;<hao.a.wu>&gt;

Got it. I will add checks before freeing those two pointers to eliminate the chance of freeing NULL pointers.

**Comment by tianocore-issues** [comment]
## Comment 631
Date: 2016-10-19 04:13:39 +0000
From: Hao Wu &lt;<hao.a.wu>&gt;

I am able to reproduce the ASSERT issue on UDK2015 branch (both HEAD version using VS2015x86 tool-chain and version Git-59fd15c5a2174a6a6fc4133964e6d5b520b58663 (i.e. svn-r18552) using VS2013x86 tool-chain) by changing the PCD &apos;gEfiNt32PkgTokenSpaceGuid.PcdWinNtFileSystem&apos; in file Nt32Pkg\Nt32Pkg.dsc to point to a non-existing directory:

Like from L&quot;.!..\..\..\..\EdkShellBinPkg\Bin\Ia32\Apps&quot; -&gt; L&quot;.!..\..\..\..\EdkShellBinPkg\Bin\Ia32\AppsTest&quot;.

Meanwhile, I notice that on edk2 master, this issue has already been fixed by commit bbd6b0106e9e725853d12550527aa03f11c09fcf. So the reported issue does not exist on master branch.

The fix on master branch is the same with the one provided by Eric Nelson, so I will apply the fix on the UDK2015 branch.

I changed both the fields &apos;Release(s) the issue is observed&apos; and &apos;Release(s) the issues must be fixed&apos; to contain only UDK2015 branch, since the reported issue only exists on this branch at the moment.

**Comment by tianocore-issues** [comment]
## Comment 648
Date: 2016-10-19 22:14:13 +0000
From: Hao Wu &lt;<hao.a.wu>&gt;

Fixed this issue on UDK2015 branch via commit:
55f0ddb2b6f533177ec6b72fd42bd355f13c61a8
